### ISSUE TITLE
BUG Report: Downtime and Comments not able to parse livestatus output

### ISSUE THREAD

**Issue by tjyang** [opinkerfi/adagios#643]
* WHAT :
```
InvalidResponseFromLivestatus: Could not parse response from livestatus. \
Query:GET downtimes ResponseHeader: fixed16 OutputFormat: python ColumnHeaders: on Response: 
```
* HOW: query on  Aadagio downtime.
* Environment info
```
CentOS 7.4 latest
adagios 1.6.3.
Nagios 4.3.4
check-mk-livestatus-1.2.8
```
* [The screenshot on my test nagios server](https://user-images.githubusercontent.com/378638/37875095-456a958c-3000-11e8-9bd2-fa66ced6b8bc.png)

* Traceback log

```Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/adagios/views.py", line 43, in wrapper
    result = view_func(request, *args, **kwargs)
  File "/usr/lib/python2.7/site-packages/adagios/status/views.py", line 971, in downtime_list
    c['downtimes'] = l.query('GET downtimes', *args)
  File "/usr/lib/python2.7/site-packages/pynag/Parsers/multisite.py", line 80, in query
    query_result = backend_instance.query(query, *args, **kwargs)
  File "/usr/lib/python2.7/site-packages/pynag/Parsers/livestatus.py", line 996, in query
    raise InvalidResponseFromLivestatus(query=livestatus_query, response=response_data)
InvalidResponseFromLivestatus: Could not parse response from livestatus.
Query:GET downtimes
ResponseHeader: fixed16
OutputFormat: python
ColumnHeaders: on


Response: [[u"author",u"comment",u"duration",u"end_time",u"entry_time",u"fixed",u"host_accept_passive_checks",u"host_acknowledged",u"host_acknowledgement_type",u"host_action_url",u"host_action_url_expanded",u"host_active_checks_enabled",u"host_address",u"host_alias",u"host_check_command",u"host_check_command_expanded",u"host_check_flapping_recovery_notification",u"host_check_freshness",u"host_check_interval",u"host_check_options",u"host_check_period",u"host_check_type",u"host_checks_enabled",u"host_childs",u"host_comments",u"host_comments_with_extra_info",u"host_comments_with_info",u"host_contact_groups",u"host_contacts",u"host_current_attempt",u"host_current_notification_number",u"host_custom_variable_names",u"host_custom_variable_values",u"host_custom_variables",u"host_display_name",u"host_downtimes",u"host_downtimes_with_info",u"host_event_handler",u"host_event_handler_enabled",u"host_execution_time",u"host_filename",u"host_first_notification_delay",u"host_flap_detection_enabled",u"host_groups",u"host_hard_state",u"host_has_been_checked",u"host_high_flap_threshold",u"host_icon_image",u"host_icon_image_alt",u"host_icon_image_expanded",u"host_in_check_period",u"host_in_notification_period",u"host_in_service_period",u"host_initial_state",u"host_is_executing",u"host_is_flapping",u"host_last_check",u"host_last_hard_state",u"host_last_hard_state_change",u"host_last_notification",u"host_last_state",u"host_last_state_change",u"host_last_time_down",u"host_last_time_unreachable",u"host_last_time_up",u"host_latency",u"host_long_plugin_output",u"host_low_flap_threshold",u"host_max_check_attempts",u"host_metrics",u"host_mk_inventory",u"host_mk_inventory_gz",u"host_mk_inventory_last",u"host_modified_attributes",u"host_modified_attributes_list",u"host_name",u"host_next_check",u"host_next_notification",u"host_no_more_notifications",u"host_notes",u"host_notes_expanded",u"host_notes_url",u"host_notes_url_expanded",u"host_notification_interval",u"host_notification_period",u"host_notifications_enabled",u"host_num_services",u"host_num_services_crit",u"host_num_services_hard_crit",u"host_num_services_hard_ok",u"host_num_services_hard_unknown",u"host_num_services_hard_warn",u"host_num_services_ok",u"host_num_services_pending",u"host_num_services_unknown",u"host_num_services_warn",u"host_obsess_over_host",u"host_parents",u"host_pending_flex_downtime",u"host_percent_state_change",u"host_perf_data",u"host_plugin_output",u"host_pnpgraph_present",u"host_process_performance_data",u"host_retry_interval",u"host_scheduled_downtime_depth",u"host_service_period"
```

**Comment by Mjolinir** [comment]
I can confirm this issue

Centos 7.4 (and Centos 7.5)
adagios-1.6.3-1
nagios-4.3.4-5
check-mk-livestatus-1.2.8p26-1

`Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/adagios/views.py", line 43, in wrapper
    result = view_func(request, *args, **kwargs)
  File "/usr/lib/python2.7/site-packages/adagios/status/views.py", line 971, in downtime_list
    c['downtimes'] = l.query('GET downtimes', *args)
  File "/usr/lib/python2.7/site-packages/pynag/Parsers/multisite.py", line 80, in query
    query_result = backend_instance.query(query, *args, **kwargs)
  File "/usr/lib/python2.7/site-packages/pynag/Parsers/livestatus.py", line 996, in query
    raise InvalidResponseFromLivestatus(query=livestatus_query, response=response_data)
InvalidResponseFromLivestatus: Could not parse response from livestatus.
Query:GET downtimes
ResponseHeader: fixed16
OutputFormat: python
ColumnHeaders: `on``

**Comment by gardart** [comment]
thank you @tjyang and @Mjolinir 
what version of Pynag are you using?

**Comment by Mjolinir** [comment]
For me it looks to be:
pynag-0.9.1-1

**Comment by gardart** [comment]
@Mjolinir and @tjyang could you try to update to the latest pynag and adagios (released last week), using 
yum --enablerepo=ok-testing update pynag adagios
let me know if this solves this issue

**Comment by tjyang** [comment]
@Mjolinir , I updated the new rpms on my test nagios instance, it didn't help. Can you confirm ?

```
After "yum --enablerepo=ok-testing update pynag adagios"
[me@nagios03 ~]$ rpm -qa |egrep 'adagio|pynag'
pynag-0.9.1-1.git.187.9bcf9ed.el7.noarch
adagios-1.6.3-2.git.0.4290a53.el7.noarch
[me@ilclnagios03 ~]$

```

**Comment by gardart** [comment]
@tjyang 
could you try to add this to your livestatus broker in /etc/nagios/nagios.cfg
debug=1 query_timeout=0

**Comment by Mjolinir** [comment]
Looks very similar for me:
`tail -50 /var/log/nagios/livestatus.log
2018-05-24 07:54:14 Time to process request: 22 us. Size of answer: 8 bytes
2018-05-24 07:54:14 Query: GET hosts
2018-05-24 07:54:14 Query: ResponseHeader: fixed16
2018-05-24 07:54:14 Query: OutputFormat: python
2018-05-24 07:54:14 Query: ColumnHeaders: on
2018-05-24 07:54:14 Time to process request: 6393 us. Size of answer: 164443 bytes
2018-05-24 07:54:14 Query: GET hosts
2018-05-24 07:54:14 Query: ResponseHeader: fixed16
2018-05-24 07:54:14 Query: OutputFormat: python
2018-05-24 07:54:14 Query: ColumnHeaders: on
2018-05-24 07:54:14 Time to process request: 6409 us. Size of answer: 164443 bytes
2018-05-24 07:54:14 Query: GET services
2018-05-24 07:54:14 Query: Filter: state != 0
2018-05-24 07:54:14 Query: Filter: acknowledged = 0
2018-05-24 07:54:14 Query: Filter: host_acknowledged = 0
2018-05-24 07:54:14 Query: Filter: scheduled_downtime_depth = 0
2018-05-24 07:54:14 Query: Filter: host_scheduled_downtime_depth = 0
2018-05-24 07:54:14 Query: Stats: state != 0
2018-05-24 07:54:14 Query: Stats: host_state != 0
2018-05-24 07:54:14 Query: ResponseHeader: fixed16
2018-05-24 07:54:14 Query: OutputFormat: python
2018-05-24 07:54:14 Query: ColumnHeaders: off
2018-05-24 07:54:14 Time to process request: 88 us. Size of answer: 8 bytes
2018-05-24 07:54:14 Query: GET services
2018-05-24 07:54:14 Query: Stats: state != 0
2018-05-24 07:54:14 Query: Stats: state != 0
2018-05-24 07:54:14 Query: Stats: acknowledged = 0
2018-05-24 07:54:14 Query: Stats: scheduled_downtime_depth = 0
2018-05-24 07:54:14 Query: Stats: host_state = 0
2018-05-24 07:54:14 Query: StatsAnd: 4
2018-05-24 07:54:14 Query: ResponseHeader: fixed16
2018-05-24 07:54:14 Query: OutputFormat: python
2018-05-24 07:54:14 Query: ColumnHeaders: off
2018-05-24 07:54:14 Time to process request: 54 us. Size of answer: 8 bytes
2018-05-24 07:54:14 Query: GET hosts
2018-05-24 07:54:14 Query: Stats: state != 0
2018-05-24 07:54:14 Query: Stats: state != 0
2018-05-24 07:54:14 Query: Stats: acknowledged = 0
2018-05-24 07:54:14 Query: Stats: scheduled_downtime_depth = 0
2018-05-24 07:54:14 Query: Stats: host_state = 1
2018-05-24 07:54:14 Query: StatsAnd: 4
2018-05-24 07:54:14 Query: ResponseHeader: fixed16
2018-05-24 07:54:14 Query: OutputFormat: python
2018-05-24 07:54:14 Query: ColumnHeaders: off
2018-05-24 07:54:14 Time to process request: 51 us. Size of answer: 8 bytes
2018-05-24 07:54:14 Query: GET hosts
2018-05-24 07:54:14 Query: ResponseHeader: fixed16
2018-05-24 07:54:14 Query: OutputFormat: python
2018-05-24 07:54:14 Query: ColumnHeaders: on
2018-05-24 07:54:14 Time to process request: 6700 us. Size of answer: 164443 bytes`

**Comment by Mjolinir** [comment]
One thing I noticed, not sure if it is relevant,

Im using check-mk-livestatus-1.2.8p26-1.el7 from EPEL.
If I use mk-livestatus-1.2.2-3.git.2.27fc0fd.el7.centos.x86_64 from ok-testing then livestatus does not work at all.

**Comment by gardart** [comment]
check-mk-livestatus-1.2.8p26-1.el7 from EPEL is the correct one...

**Comment by tjyang** [comment]
@gardart 
I am using  ***check-mk-livestatus-1.2.8p26-1.el7 from EPEL*** , Comments and Downtime still has issue. Looks like the adagios side of parser code need to be adjusted.

**Comment by gardart** [comment]
does your nagios server crash when this happens? Do you need to restart nagios service every time?

**Comment by tjyang** [comment]
No, both nagios and livestatus  daemon weren't not crashed when  this issue happened. 

```
[root@nagios03 nagios]# tail -20f  /var/log/nagios/livestatus.log
2018-07-11 16:32:19 Idle timeout of 12000 ms exceeded. Going to close connection.
2018-07-11 16:32:19 error: Client connection terminated while request still incomplete
2018-07-11 16:32:21 Idle timeout of 12000 ms exceeded. Going to close connection.
2018-07-11 16:32:21 error: Client connection terminated while request still incomplete
2018-07-11 16:32:41 Idle timeout of 12000 ms exceeded. Going to close connection.
2018-07-11 16:32:41 error: Client connection terminated while request still incomplete
2018-07-11 16:32:48 Idle timeout of 12000 ms exceeded. Going to close connection.
2018-07-11 16:32:48 error: Client connection terminated while request still incomplete
2018-07-11 20:01:04 deinitializing
2018-07-11 20:01:04 Waiting for main to terminate...
2018-07-11 20:01:04 Waiting for client threads to terminate...
2018-07-11 20:01:04 Logfile cache: flushing complete cache.
2018-07-12 00:01:03 deinitializing
2018-07-12 00:01:03 Waiting for main to terminate...
2018-07-12 00:01:05 Waiting for client threads to terminate...
2018-07-12 00:01:05 Logfile cache: flushing complete cache.
2018-07-12 04:01:04 deinitializing
2018-07-12 04:01:04 Waiting for main to terminate...
2018-07-12 04:01:06 Waiting for client threads to terminate...
2018-07-12 04:01:06 Logfile cache: flushing complete cache.
^C
[root@ilclnagios03 nagios]# date
Thu Jul 12 07:01:13 EDT 2018
[root@nagios03 nagios]#

```

**Comment by Mjolinir** [comment]
Same applies to me. no crashes.

**Comment by Mjolinir** [comment]
I noticed today that both Comments and Downtime are working! Unfortunately I am not sure which update fixed it. Here are the current versions of related packages:

check-mk-livestatus-1.4.0p31-2.el7.x86_64 (last updated June 21)
pynag-0.9.1-1.git.187.9bcf9ed.el7.noarch (last updated May 24)
adagios-1.6.3-2.git.0.4290a53.el7.noarch (last updated May 24)
nagios-4.3.4-5.el7.x86_64 (last updated Apr 16)

It seems likely it was the check-mk-livestatus update in June and I just didn't notice - the updates are automated with Ansible

@tjyang can you confirm on your end?

**Comment by tjyang** [comment]
- Existing rpms
```
[me@nagios03 ~]$  rpm -qa |egrep 'check-mk-livestatus-1|pynag-0|adagios-1|nagios-4'
pynag-0.9.1-1.git.187.9bcf9ed.el7.noarch
adagios-1.6.3-2.git.0.4290a53.el7.noarch
nagios-4.3.4-3.el7.x86_64
check-mk-livestatus-1.2.8p26-1.el7.x86_64
[me@nagios03 ~]$
```
* update livestatus
```
sudo  yum update -y check-mk-livestatus
sudo systemctl restart nagios
```
* This livestatus fixed the "Comment" and "Downtimes" menu. ;)

Thanks to @Mjolinir's pointer and @gardart's help.

**Comment by tjyang** [comment]
* Recording another running instances on CentOS 7.6 without showing comment issue.

```
[root@nagios03 ~]# cat /etc/redhat-release; rpm -qa |egrep 'check-mk-livestatus-1|pynag-0|adagios-1|nagios-4';date
CentOS Linux release 7.6.1810 (Core)
pynag-0.9.1-1.git.187.9bcf9ed.el7.noarch
adagios-1.6.3-2.git.0.4290a53.el7.noarch
check-mk-livestatus-1.4.0p31-2.el7.x86_64
nagios-4.4.3-1.el7.x86_64
Wed Sep  4 15:17:30 EDT 2019
[root@nagios03 ~]#

```

**Comment by tjyang** [comment]
@gardart 
check-mk-livestatus-1.4.0p31-2.el7.x86_64 fixed my comment/downtime display issue but it will crash my nagios server due to livestatus aborted when doing LQL 'GET hosts' command.
I tried compiling version from 1.2.8 up to latest 1.6 , they all crashed nagios server when doing GET hosts.
so I followed  your tip above, using version 1.2.6 and now both 'GET hosts' and "comment/downtime" all works.  Thanks again for your pointer.
